[mono][jit] Fix and optimize llvmonly+gsharedvt code generation

vargaz opened the pull request.

vargaz committed:
[mono][jit] Revert parts of 62eafc1e67d332b902772065cb36b3c51c40c732.

The new approach doesn't work because it requires gsharedvt out
wrappers with signatures which might not exist in the program.

vargaz committed:
Fix interp->aot calls to gsharedvt methods, they don't need a gsharedvt out wrapper.

vargaz committed:
Fix the initialization of info->invoke_impl in mini_llvmonly_init_delegate (). Also initialize del->method there instead of doing it in generated code.

vargaz committed:
Fix interp entry for gsharedvt methods.

vargaz committed:
Extract vtable.interp_vtable into a MonoVTableEEData structure so more per-vtable ee data can be added later.

vargaz committed:
Enable gsharedvt on wasm as an experiment.

vargaz committed:
Add a fastpath for gsharedvt virtual calls.

vargaz committed:
Fix/optimize interp entries for methods which cannot be AOTed.

* Instead of obtaining the interp entry wrapper using a JIT icall,
  store it as an AOT constant/rgctx entry.
* For gsharedvt methods, collect the arguments and call into
  the interpreter directly instead of using a gsharedvt out
  + interp entry wrapper.

BrzVlad commented on src/mono/mono/mini/llvmonly-runtime.c:
```diff
@@ -215,6 +215,29 @@ mini_llvmonly_add_method_wrappers (MonoMethod *m, gpointer compiled_method, gboo
+}
+
+static void
+alloc_gsharedvt_vtable (MonoVTable *vtable)
```
I think there is some concurrency fishiness here. Callers of this method probably expect `vtable->ee_data->gsharedvt_vtable` to have been created, however it seems possible that in the meantime the interpreter could just replace the `vtable->ee_data` resulting in a NULL gsharedvt table.

vargaz committed:
[mono][jit] Avoid computing a vtable arg for ctors which don't need it.

This can happen if the ctor is implemented as an intrinsics like
System.ByReference ().

vargaz committed:
[mono][jit] Allow inline rgctx fetches during some parts of the decompose pass.

vargaz committed:
[mono][jit] Add more specialized imt functions in llvmonly mode.

vargaz committed:
[mono][aot] Fix the addition of generic methods from profiles.

Same as 0666ebc475871c27f5b9d4ee8e91922f20be46e9 but for generic methods.

vargaz committed:
Disable gsharedvt on wasm.

lambdageek commented on src/mono/mono/mini/llvmonly-runtime.c:
```diff
@@ -215,6 +215,29 @@ mini_llvmonly_add_method_wrappers (MonoMethod *m, gpointer compiled_method, gboo
+		mono_memory_barrier ();
+		vtable->ee_data = ee_data;
+	}
+	return ee_data;
```
Similar to what @BrzVlad said: isn't there still a race condition here?  Two threads can both call `m_class_alloc0` and both return their own freshly-allocated ee_data.  

Also, you have 2 copies of this function. here and also in `interp.c`
```suggestion
	return ee_data;
```
```suggestion
	return ee_data;
```

vargaz replied:
Will fix that in a follow up PR.
